rest-client/src/main/java/org/sagebionetworks/bridge/rest/AuthenticationHandler.java

```diff
@@ -1,96 +1,73 @@
-     * @param userSessionInfoProvider
-     *         for retrieving the user's session
-     */
-    public AuthenticationHandler(SignIn signIn, UserSessionInfoProvider userSessionInfoProvider) {
-        Preconditions.checkNotNull(signIn);
-        Preconditions.checkNotNull(userSessionInfoProvider);
-        this.signIn = signIn;
+    public AuthenticationHandler(UserSessionInfoProvider userSessionInfoProvider) {
+        checkNotNull(userSessionInfoProvider);
         this.userSessionInfoProvider = userSessionInfoProvider;
     }
-
-
+    
     @Override
-    public okhttp3.Request authenticate(Route route, okhttp3.Response response) throws IOException {
-        // if we reach this part of the code, the server had returned a 401 and userSession is
-        // invalid
-        userSessionInfoProvider.purgeCachedSession(this.userSession);
-        this.userSession = null;
-
-        if (tryCount >= MAX_TRIES || !isValidSignIn(signIn)) {
-            LOG.info("Maximum retries reached");
-            this.tryCount = 0;
-
-            // returning null terminates the request chain and propagates the 401
-            return null;
+    public Request authenticate(Route route, Response response) throws IOException {
+        // We received a 401 from the server... attempt to reauthenticate.
+        if (requiresAuth(response.request())) {
+            userSessionInfoProvider.reauthenticate();    
+            // We should now be able to proceed with session headers.
+            return addBridgeHeaders(response.request());
         }
-        tryCount++;
-
-        userSession = userSessionInfoProvider.retrieveSession(signIn);
-
-        // interceptor was already triggered for this request and failed to authenticate
-        // add headers again, now that we've retrieved a session again
-        return addBridgeHeaders(response.request());
-    }
-
-    private boolean isValidSignIn(SignIn signIn) {
-        return signIn != null && !Strings.isNullOrEmpty(signIn.getEmail()) && !Strings
-                .isNullOrEmpty(signIn.getPassword());
+        return null;
     }
 
     @Override
-    public okhttp3.Response intercept(Chain chain) throws IOException {
-        okhttp3.Response response = chain.proceed(addBridgeHeaders(chain.request()));
-
-        // if we reach this part of the code, we didn't get a 401 and the authenticator did its job
-        this.tryCount = 0;
-
-        return response;
+    public Response intercept(Chain chain) throws IOException {
+        Request request = chain.request();
+        
+        // Basically, everything in the auth controller can skip a session token, except for sign out
+        if (requiresAuth(request)) {
+            request = addBridgeHeaders(request);
+        }
+        return chain.proceed(request);
+    }
+    
+    private boolean requiresAuth(Request request) {
+        String url = request.url().toString();
+        return (!url.contains(AUTH_PATH) || url.endsWith(SIGN_OUT_PATH));
```

Comment:
this looks like it would return requiresAuth=true for signout

Reply:
Yes, you need to send us the session token or we can sign you out. We don't know who you are.

Reply:
Totally makes sense we need to addBridgeHeader for signIn and signOut. But do we want to reauthenticate before calling sign out? I don't think it hurts, and IIRC this would invalidate the session itself and effectively sign you out of other devices.

Reply:
If we want to reauth before sign out, a comment would help explain why